### PR TITLE
[refactor] Remove the singleton Program::context

### PR DESCRIPTION
**Pull request by k-ye**
Finally..

Related issue = #1739

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/transforms/constant_fold.cpp [hunk 1]
```diff
@@ -137,12 +135,10 @@ class ConstantFold : public BasicStmtVisitor {
                       stmt->cast_type,
                       false};
     auto *ker = get_jit_evaluator_kernel(id);
-    auto &current_program = stmt->get_kernel()->program;
-    // save input args to prevent the current kernel from being overridden.
-    ContextArgSaveGuard _(current_program.get_context());
```
Comment (archibate) [inline taichi/transforms/constant_fold.cpp:142]: Delete `class ContextArgSaveGuard` from this file too?

**Comment by archibate** [comment]
Great, just one question, will we reuse the new `args[0]` as return value iapr?

File: taichi/transforms/constant_fold.cpp (hunk 1 above)
Comment (k-ye) [reply taichi/transforms/constant_fold.cpp:142]: Done

**Comment by k-ye** [comment]
> will we reuse the new args[0] as return value iapr?

Not in the near future. All the backends are now copying the result back to `result_buffer`. These contexts are temporary objects now, so there can be no reference to them after a kernel launch.

**Comment by archibate** [comment]
> there can be no reference to them

Why we need reference? Wasn't `result_buffer` actually preventing `Program` from async?

**Comment by k-ye** [comment]
> Why we need reference?

I mean we cannot refer to these `Context` objects once a kernel has launched (after `Kernel::operator()`).. 

> Wasn't result_buffer actually preventing Program from async?

How so?

File: taichi/transforms/constant_fold.cpp [hunk 2]
```diff
@@ -114,13 +100,11 @@ class ConstantFold : public BasicStmtVisitor {
...
     auto launch_ctx = ker->make_launch_context();
     launch_ctx.set_arg_raw(0, lhs.val_u64);
     launch_ctx.set_arg_raw(1, rhs.val_u64);
     (*ker)(launch_ctx);
+    auto &current_program = stmt->get_kernel()->program;
     ret.val_i64 = current_program.fetch_result<int64_t>(0);
```
Comment (archibate) [inline taichi/transforms/constant_fold.cpp:108]: My future hope is:
```suggestion
    ret.val_i64 = launch_ctx.get_arg<int64_t>(0);
```

**Review by archibate — APPROVED** [review]
LGTM now! What else can I do for opengl & cc later?

File: taichi/transforms/constant_fold.cpp (hunk 2 above)
Comment (k-ye) [reply taichi/transforms/constant_fold.cpp:108]: I see your point. Unfortunately, reading the result is, in most cases, separated from calling the kernel.. I guess a preliminary step toward this goal would be to stop using `Program` to fetch the result.

**Comment by k-ye** [comment]
> What else can I do for opengl & cc later?

I think you've already covered opengl and cc :) The major goal of this cleanup is to remove the singleton context from `Program`. Before the cleanup, this has quite complicated the CUDA logic..

**Comment by archibate** [comment]
> > What else can I do for opengl & cc later?
> 
> I think you've already covered opengl and cc :) The major goal of this cleanup is to remove the singleton context from `Program`. Before the cleanup, this has quite complicated the CUDA logic..

I see, so would you (and @yuanming-hu) share more details about the `async_mode`? I'd happy to help with.

**Comment by yuanming-hu** [comment]
> > > What else can I do for opengl & cc later?
> > 
> > 
> > I think you've already covered opengl and cc :) The major goal of this cleanup is to remove the singleton context from `Program`. Before the cleanup, this has quite complicated the CUDA logic..
> 
> I see, so would you (and @yuanming-hu) share more details about the `async_mode`? I'd happy to help with.

That's a highly premature research project. We will document it once we finish prototyping and after the related paper is published, hopefully in early 2021.

**Pull request merged**